File: src/tomek.js/compiler/template/text_expression.rb
# To change this template, choose Tools | Templates
# and open the template in the editor.

require 'json'

class TextExpression
 
	class JSString < String
		
	end
	
	class JSExpression < String
		
	end
	
	@str
	
	def initialize str
		parts = []
		
		mode = "text"
		part = ""
		
		i = 0
		str = str.to_s.strip
		l = str.length
		
		while i < ( l + 1 ) do
		
			if mode == "text" and ( i >= l or str[i,3] == "[%=" ) then
				mode = "expr"
				i += 3
				
				part = part.strip
				if part.length > 0 then
					part = part.to_json
					parts.push part
				end
				
				part = ""
			end
			
			if mode == "expr" and ( i >= l or str[i,2] == "%]" ) then
				mode = "text"
				i += 2
				
				part = part.strip
				if part.length > 0 then
					part = "( new TExpression( function(){ return ("+part+"); }.bind( ExpressionContext ) ) )"
					parts.push part
				end
				
				part = ""
			end
			
			if ( i < l ) then
				part += str[i,1]
			end
			
			i += 1
		end

    @str = parts.join( "+" )
  end
	
	def to_s
		@str
	end
	
	def js_expression
		@str
	end
	
	def escape_js_string str
		str.to_json
	end
	
	
	
end

File: src/tomek.js/compiler.js/template/TextExpression.js
var TextExpression = (function () {
    function TextExpression(expr) {
        this.epression = this.createExpressionString(expr);
    }
    TextExpression.prototype.createExpressionString = function (str) {
        var parts = [];
        var l = str.length;
        var i = 0;
        var part = "";
        var mode = "text";
        while (i < (l + 1)) {
            if (mode == "text" && str.substring(i, i + 3) == "[%@") {
                mode = "trans";
                i += 3;
                if (part.length > 0) {
                    parts.push(JSON.stringify(part));
                }
                part = "";
            }
            if (mode == "text" && str.substring(i, i + 3) == "[%=") {
                mode = "expr";
                i += 3;
                if (part.length > 0) {
                    parts.push(JSON.stringify(part));
                }
                part = "";
            }
            if (mode == "trans" && str.substring(i, i + 2) == "%]") {
                mode = "text";
                i += 2;
                if (part.length > 0) {
                    //@TODO
                    var translation = "";
                    parts.push(JSON.stringify(translation));
                }
                part = "";
            }
            if (mode == "expr" && str.substring(i, i + 2) == "%]") {
                mode = "text";
                i += 2;
                if (part.length > 0) {
                    //@TODO
                    var expr_part = "( new TExpression( function(){ return (" + part + "); }.bind( ExpressionContext ) ) )";
                    parts.push(expr_part);
                }
                part = "";
            }
            if (i < l) {
                part += str.substring(i, i + 1);
            }
            i += 1;
        }
        if (parts.length <= 0) {
            parts.push("");
        }
        return parts.join("+");
    };
    return TextExpression;
})();
exports.__esModule = true;
exports["default"] = TextExpression;
